k9mail/src/main/java/com/fsck/k9/message/html/HttpUriParser.java

```diff
@@ -40,19 +40,15 @@ public int linkifyUri(String text, int startPos, StringBuffer outputBuffer) {
         }
 
         // Authority
-        int authorityEnd = text.indexOf('/', currentPos);
-        if (authorityEnd == -1) {
-            authorityEnd = text.length();
-        }
-
-        currentPos = matchUserInfoIfAvailable(text, currentPos, authorityEnd);
+        currentPos = matchUserInfoIfAvailable(text, currentPos);
 
-        if (!tryMatchDomainName(text, currentPos, authorityEnd) &&
-                !tryMatchIpv4Address(text, currentPos, authorityEnd, true) &&
-                !tryMatchIpv6Address(text, currentPos, authorityEnd)) {
+        int matchedAuthorityEnd = Math.max(tryMatchDomainName(text, currentPos),
```

Comment:
The whole `Math.max()` business makes this super hard to read. There's also no need to attempt another match if one of the methods was successful. So I suggest to extract all of this to a separate method and then to check the return value after each call to `tryMatch*()` and return early if a match was found.